### ISSUE TITLE
FNX-4962 ⁃ [Bug] History: Consecutive Duplicates should be filtered out

### ISSUE THREAD

**Issue by cadeyrn** [mozilla-mobile/fenix#1331]
### Steps to reproduce
1. open history

### Expected behavior
no duplicated entries, because the duplicates don't add much value

### Actual behavior
duplicated entries

![Screenshot](https://user-images.githubusercontent.com/1100614/55434659-5b7fc180-5598-11e9-9438-fc01ca52ba2b.png)UNITO-UNDERSCORE!20190402-223933!

### Device information
* Android device: HTC U11 / Android 8.0
* Fenix version: Fenix revision bd81e722393381458e3c1421773ee402121e9dc7

**Comment by pocmo** [comment]
Related AC issues:
* https://github.com/mozilla-mobile/android-components/issues/2606
* https://github.com/mozilla-mobile/android-components/issues/2643

**Comment by ekager** [comment]
Those 2 AC issues are merged in and being used. We should have higher quality history now.

**Comment by softvision-miralobontiu** [comment]
Hi @cadeyrn,
I'm not completely sure if I understood correctly the expected behavior. 
For example, I searched for YouTube and in History, I get http://youtube.com, and https://m.youtube.com/, although I typed "youtube" inside the tab, and I didn't select any video. Is this expected or not?
Thank you!

Please see the attached screenshot:

![Screenshot_20190529-165112](https://user-images.githubusercontent.com/40694158/58562861-c36a2580-8232-11e9-87e1-1d3a5e2b636a.png)

**Comment by ekager** [comment]
@grigoryk could you possibly weigh in on if the above screenshot is the "best" we can do with history filtering right now?

**Comment by sblatz** [comment]
Still need @grigoryk's input on this.

**Comment by sarentz** [comment]
@mhammond are these duplicates something that should be filtered/merged in the a-s code?

**Comment by jhugman** [comment]
The `SELECT` statement is coming in from [`history.rs`][1] over the [`moz_historyvisits TABLE`][2] with foreign key references to the [`moz_places TABLE`][3].

From comment #0 above it's not clear if desired behaviour is filtering duplicates (observe multiple `about:blank`) versus filtering duplicate URLs, but duplicate URLs modulo differences in scheme and PSL (`http://soeren-hetzl.com` v `https://soeren-hetzl.com` and `youtube.com` v `m.youtube.com`).

Pinging @mhammond @linacambridge who might know if this would be better/easier done in rust or kotlin).

[1]: https://github.com/mozilla/application-services/blob/8432a5a0e4a627cf0ceb7eea72a23168e99a96ac/components/places/src/storage/history.rs#L1052
[2]: https://github.com/mozilla/application-services/blob/1c9f1ba4ef0921e7eb57e89842254db8c4c511ae/components/places/sql/create_shared_schema.sql#L62
[3]: https://github.com/mozilla/application-services/blob/1c9f1ba4ef0921e7eb57e89842254db8c4c511ae/components/places/sql/create_shared_schema.sql#L9

**Comment by thomcc** [comment]
We intentionally aren't filtering duplicates, but could easily add a parameter to do so.

**Comment by thomcc** [comment]
More broadly, there are a couple things you might want here:

1. Only show the (globally) most recent visit for a given URL. This is consistent with desktop's handling, and is easy for us to do, however it would mean that deeper into your history pages, the data might be noticably less accurate.
2. Within a given page, only show the most recent visit of a specific url. This is also easy for us to add, but the difference between adding it in rust and kotlin wouldn't be huge -- we'd just be doing the filtering after the query.
3. Other variants... I could imagine also a situation where only consecutive duplication is filtered...

Basically, I think if you want number 1, we should do it in the Rust.

Otherwise, it probably is worth you doing it in Kotlin, so that you retain that flexibility to change it later without needing to go through us.

**Comment by jdragojevic** [comment]
@vesta0 - Can you please consider the options that @thomcc outlined above for behaviours and let us know how you like it to be resolved?

**Comment by yoasif** [comment]
>Other variants... I could imagine also a situation where only consecutive duplication is filtered...

I think consecutive duplication is the only one that is desirable for users. People on desktop already complain about behavior #1, and I don't see any reason why it is desirable to be consistent with this disliked behavior.

**Comment by vesta0** [comment]
@thomcc thanks for listing the options. 

I don't think we should only show a most recent visit for a URL. Comparing how this works in Fennec vs Fenix, I think the problem is that Fenix is not pulling the right page name all the time. For example, if you tap on different ideos on YouTube in Fenix they will all have the same name (m.youtube.com) even if the URL is different, whereas Fennec correctly displays the name of the video. Can we match this  behaviour in Fenix?

Also what does the effort for option 3 (filtering out only consecutive dups) look like?

**Comment by jhugman** [comment]
So, threading the needle through all the above: 

 * we should filter out duplicate consecutive entries.
 * I think we should do this in Kotlin, to give us some ongoing flexibility.

Other issues alluded to in this bug:

#### 1. title of the page is not being saved 

> Fenix is not pulling a correct/distinct page name all the time. For example, if you tap on different ideos on YouTube in Fenix they will all have the same name (m.youtube.com) even if the URL is different, whereas Fennec correctly displays the name of the video.

#### 1a. intermediate pages in a redirect should not be stored in history.

#### 2. filtering duplicates in history should be using title, or PSL/common prefixes trimmed URL.

Solving 2. by filtering by title would be relatively cheap, however, we cannot trust that the title for consecutive duplicates are actually the same, because 1. exists. Until then, we have garbage in, garbage out.

**Comment by st3fan** [comment]
This may be an unrelated problem but it is worth investigating:

> Fenix is not pulling a correct/distinct page name all the time. For example, if you tap on different ideos on YouTube in Fenix they will all have the same name (m.youtube.com) even if the URL is different, whereas Fennec correctly displays the name of the video.

Is this unique to YouTube because it reloads the page not with actual reloads but by doing ajax requests and modifying the history stack? We've had this that problem in Firefox iOS where we did not get the proper notifications that the page changed.

This is not unique to YouTube - there are plenty of SPAs that use these techniques.

**Comment by jhugman** [comment]
@st3fan `pushState` may well be relevant here.

I'll file an ac issue to track that.
